### ISSUE TITLE
Collapse `load; add; store` into a single `add` on x86-64

### ISSUE THREAD

**Issue by fitzgen** [bytecodealliance/wasmtime#1925]
Right now, this

```
function %add(i64) {
    block0(v0: i64):
        v1 = load.i32 aligned notrap v0
        v2 = iadd_imm v1, 1
        store aligned notrap v2, v1
        return
}
```

compiles down into

```
Disassembly of 19 bytes:
   0:	40 55                	push	rbp
   2:	48 89 e5             	mov	rbp, rsp
   5:	40 8b 07             	mov	eax, dword ptr [rdi]
   8:	89 c1                	mov	ecx, eax
   a:	83 c1 01             	add	ecx, 1
   d:	40 89 08             	mov	dword ptr [rax], ecx
  10:	40 5d                	pop	rbp
  12:	c3                   	ret	
```

on x86-64.

I would expect it to instead generate

```
push rbp
mov  rbp, rsp
add  dword ptr [rdi], 1
pop  rbp
ret
```

but I don't think we any IR instruction handles this addressing mode right now.

**Comment by fitzgen** [comment]
(Using `--set opt_level=speed`)

**Comment by abrown** [comment]
Which backend? I don't think the old backend can do this (due to limited addressing modes in the recipes) but I was planning to add this optimization to the new backend eventually since that has support for different addressing modes.

**Comment by abrown** [comment]
(and by "I was planning to add this" I mean feel free to do this since it will take me a bit to get there).

**Comment by fitzgen** [comment]
The current backend was all I tested.

**Comment by bjorn3** [comment]
`iadd_mem`, `isub_mem`, etc (any suggestions fir a better name) instructions would also be necessary to implement atomic increment.

**Comment by julian-seward1** [comment]
In the new backend, this would require a new `Inst` variant, `Alu_R_RM` or perhaps `Alu_RI_RM`.  That bit is easy.  More complex is that we'd have to add structure to the new instruction selection framework to match the multiple CLIR insns and produce a single x64 insn.  Currently it can do that for pure-value trees rooted at some node in the CLIR insn dag, but this doesn't fit into that scenario.  We'd have to confer with @cfallin.

**Comment by bnjbvr** [comment]
This is entirely doable in the new backend indeed, with less effort than would be required for the old one (viz., no need to add an extra specific CLIF instruction). I would strongly recommend to not implement this in the old backend at this point (more to come about this in the next few hours/days!).

**Comment by bjorn3** [comment]
> viz., no need to add an extra specific CLIF instruction

What I meant is that a specific CLIF instruction is needed anyway for atomic increment, so it could just as well have a non-atomic variant. I did imagine that atomicness would be a `MemFlag` instead of duplicating every instruction that operates on memory.

**Comment by fitzgen** [comment]
FWIW, having a specific instruction for this would be nice too. I'd like to use it in wasmtime's gc barriers for reference types.

**Comment by fitzgen** [comment]
The infrastructure for the new backend to do this collapsing landed in https://github.com/bytecodealliance/wasmtime/pull/2366 and some follow up work for merging ALU ops and memory loads in https://github.com/bytecodealliance/wasmtime/pull/2389

I guess we should decide whether we ever want to support an `iadd_imm_mem` instruction in clif itself, or if we want this to only be something that the backend optimizes risc-y instructions into. If we don't want to add such an instruction, then I think we can close this issue.

**Comment by cfallin** [comment]
I think that the latter is probably the better option; in general we want to try to reduce the instruction count going forward, and do better with pattern-matching.

This isn't too much more work on top of what we already have (load+op); we can either keep this issue open to track that work (the title seems to match what we want to do still?) or open a new one, up to you!

**Comment by fitzgen** [comment]
Sounds good. Let's keep this open to track combining load+op+store into a single instruction.

**Comment by fitzgen** [comment]
And we should probably double check that the GC barriers do end up using a single `add` in practice, too.

**Comment by cfallin** [comment]
Closing now; fixed by #4071.
